pkg/operator/ceph/nfs/controller.go

```
@@ -17,214 +17,287 @@ limitations under the License.
+	if !isReadyToReconcile {
+		// This handles the case where the Ceph Cluster is gone and we want to delete that CR
+		// We skip the deleteStore() function since everything is gone already
+		//
+		// Also, only remove the finalizer if the CephCluster is gone
+		// If not, we should wait for it to be ready
+		// This handles the case where the operator is not ready to accept Ceph command but the cluster exists
+		if !cephNFS.GetDeletionTimestamp().IsZero() && !cephClusterExists {
+			// Remove finalizer
+			err := opcontroller.RemoveFinalizer(r.client, cephNFS)
+			if err != nil {
+				return reconcile.Result{}, errors.Wrap(err, "failed to remove finalizer")
+			}
+
+			// Return and do not requeue. Successful deletion.
+			return reconcile.Result{}, nil
+		}
 
-	if !nfsChanged(oldNFS.Spec, newNFS.Spec) {
-		logger.Debugf("nfs ganesha %q not updated", newNFS.Name)
-		return
+		logger.Debugf("CephCluster resource not ready in namespace %q, retrying in %q.", request.NamespacedName.Namespace, opcontroller.WaitForRequeueIfCephClusterNotReadyAfter.String())
+		return reconcileResponse, nil
 	}
+	r.cephClusterSpec = &cephClusterSpec
 
-	c.acquireOrchestrationLock()
-	defer c.releaseOrchestrationLock()
+	// Populate clusterInfo
+	// Always populate it during each reconcile
+	clusterInfo, _, _, err := mon.LoadClusterInfo(r.context, request.NamespacedName.Namespace)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to populate cluster info")
+	}
+	r.clusterInfo = clusterInfo
 
-	logger.Infof("Updating the ganesha server from %d to %d active count", oldNFS.Spec.Server.Active, newNFS.Spec.Server.Active)
+	// Populate CephVersion
+	daemon := string(opconfig.MonType)
+	currentCephVersion, err := cephclient.LeastUptodateDaemonVersion(r.context, r.clusterInfo.Name, daemon)
+	if err != nil {
+		logger.Errorf("failed to retrieve current ceph %q version. %v", daemon, err)
```

we don't need to return an error here?